### Exploring used cars data from the German ebay website

##### Here we will be exploring data about used cars from the German ebay website.

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv("autos.csv", encoding="latin1")

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

##### Dataset has 50,000 rows. Columns 'vehicle type', 'gearbox', 'model', 'fueltype', 'notrepaireddamange' have some null values. All columns are either strings or 64 bit integers.

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
new_col_names = {'dateCrawled': 'date_crawled', 'name': 'name', 'seller': 'seller', 'offerType': 'offer_type', 'price': 'price', 'abtest': 'ab_test',
       'vehicleType': 'vehicle_type', 'yearOfRegistration': 'registration_year', 'gearbox': 'gearbox', 'powerPS': 'power_PS', 'model': 'model',
       'odometer': 'odometer', 'monthOfRegistration': 'registration_month', 'fuelType': 'fuel_type', 'brand': 'brand',
       'notRepairedDamage': 'unrepaired_damange', 'dateCreated': 'ad_created', 'nrOfPictures': 'nr_of_pictures', 'postalCode': 'postal_code',
       'lastSeen': 'last_seen'}

In [7]:
autos = autos.rename(columns=new_col_names)

In [8]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damange           ad_created  nr_of_pictures  \
0     peugeot               nein  2016-03-26 00:00:00               0   
1         bmw               nein  2016-04-04 00:00:00               0   
2  volkswagen               nein  2016-03-26 00:00:00               0   
3       smart               nein  2016-03-12 00:00:00               0   
4        ford               nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

##### Changed camelcase to snakecase, the standard naming convention for python. 

##### Note, in line 21, I had to assign the renamed dataframe back to the autos variable as when I called the autos variable, I didnt show the renamed dataframe.

In [9]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-19 17:36:18  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand  \
count       50000        50000.000000     45518       50000   
unique         13                 NaN         7          40   
top     150,000km                 NaN    benzin  volkswagen   
freq        32424                 NaN     30107       10687   
mean          NaN            5.723360       NaN         NaN   
std           NaN            3.711984       NaN         NaN   
min           NaN            0.000000       NaN         NaN   
25%           NaN            3.000000       NaN         NaN   
50%           NaN            6.000000       NaN         NaN   
75%           NaN            9.000000       NaN         NaN   
max           NaN           12.000000       NaN         NaN   

       unrepaired_damange           ad_created  nr_of_pictures   postal_code  \
count               40171                50000         50000.0  50000.000000   
unique                  2                   76             NaN           NaN   
top                  nein  2016-04-03 00:00:00             NaN           NaN   
freq                35232                 1946             NaN           NaN   
mean                  NaN                  NaN             0.0  50813.627300   
std                   NaN                  NaN             0.0  25779.747957   
min                   NaN                  NaN             0.0   1067.000000   
25%                   NaN                  NaN             0.0  30451.000000   
50%                   NaN                  NaN             0.0  49577.000000   
75%                   NaN                  NaN             0.0  71540.000000   
max                   NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

##### Seller and offer_type have only 2 unique values one of which appears 4999 times, so these columns can be dropped.

##### Cols that may need further investigating: resgistration_year and registration_month because it is showing NaN for top and unique values and hence unexpected min/max stats. Nr_of_pictures may also need investigating because that column has no stats.

##### Columns price and odometer need to be converted to numeric.

In [10]:
##### Investigate registation_year
np.sort(autos["registration_year"].unique())

array([1000, 1001, 1111, 1500, 1800, 1910, 1927, 1929, 1931, 1934, 1937,
       1938, 1939, 1941, 1943, 1948, 1950, 1951, 1952, 1953, 1954, 1955,
       1956, 1957, 1958, 1959, 1960, 1961, 1962, 1963, 1964, 1965, 1966,
       1967, 1968, 1969, 1970, 1971, 1972, 1973, 1974, 1975, 1976, 1977,
       1978, 1979, 1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988,
       1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999,
       2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010,
       2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2800, 4100,
       4500, 4800, 5000, 5911, 6200, 8888, 9000, 9996, 9999])

In [11]:
##### Investigate registation_month
np.sort(autos["registration_month"].unique())

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12])

##### Registration_year column has years that don't make sense e.g. 

In [12]:
# Set print options
np.set_printoptions(threshold=np.inf)

In [13]:
# Clean price column and convert to float
autos["price"] = autos["price"].str.replace("$", "")
autos["price"] = autos["price"].str.replace(",", "")
autos["price"] = autos["price"].astype("float")
autos["price"].dtype

dtype('float64')

In [14]:
autos["price"].dtype

dtype('float64')

In [15]:
autos["odometer"] = autos["odometer"].str.replace("km", "")
autos["odometer"] = autos["odometer"].str.replace(",", "")
autos["odometer"] = autos["odometer"].astype("float")
autos["odometer"].dtype

dtype('float64')

In [16]:
autos = autos.rename(columns={"odometer":"odometer_km"})
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  5000.0  control          bus               2004   
1  privat    Angebot  8500.0  control    limousine               1997   
2  privat    Angebot  8990.0     test    limousine               2009   
3  privat    Angebot  4350.0  control   kleinwagen               2007   
4  privat    Angebot  1350.0     test        kombi               2003   

     gearbox  power_PS   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere     150000.0                   3       lpg   
1  automatik       286     7er     150000.0                   6    benzin   
2    manuell       102    golf      70000.0                   7    benzin   
3  automatik        71  fortwo      70000.0                   6    benzin   
4    manuell         0   focus     150000.0                   7    benzin   

        brand unrepaired_damange           ad_created  nr_of_pictures  \
0     peugeot               nein  2016-03-26 00:00:00               0   
1         bmw               nein  2016-04-04 00:00:00               0   
2  volkswagen               nein  2016-03-26 00:00:00               0   
3       smart               nein  2016-03-12 00:00:00               0   
4        ford               nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [17]:
# Explore odometer_km
autos["odometer_km"].unique().shape

(13,)

In [18]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [19]:
autos["odometer_km"].value_counts()

odometer_km
150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
5000.0        967
40000.0       819
30000.0       789
20000.0       784
10000.0       264
Name: count, dtype: int64

In [20]:
# Explore price data
autos["price"].unique().shape

(2357,)

In [21]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [22]:
autos["price"].value_counts().sort_index(ascending=True).head(10)

price
0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
Name: count, dtype: int64

In [23]:
autos_no_outliers = autos[autos["price"].between(1, 28000000)]

In [24]:
autos_no_outliers["price"].describe()

count    4.857800e+04
mean     8.069542e+03
std      1.800356e+05
min      1.000000e+00
25%      1.200000e+03
50%      3.000000e+03
75%      7.490000e+03
max      2.732222e+07
Name: price, dtype: float64

##### I have now removed the outlier of 99,999,999 and all the null values in the price column and assigned the new dataframe to a new variable. The remaining values in the price column have a mean of 8,070.

In [25]:
# Distribution of date_crawled
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

date_crawled
2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: proportion, dtype: float64

##### The date crawled is between 5th March and 7th April of 2016 with approximately 3% of cars crawled each day.

In [26]:
# Set print options
pd.set_option('display.max_columns', None)  # or 1000
pd.set_option('display.max_rows', None)  # or 1000
pd.set_option('display.max_colwidth', None)  # or 199

In [27]:
# Distribution of ad_created
autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

ad_created
2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
2015-12-30    0.00002
2016-01-03    0.00002
2016-01-07    0.00002
2016-01-10    0.00004
2016-01-13    0.00002
2016-01-14    0.00002
2016-01-16    0.00002
2016-01-22    0.00002
2016-01-27    0.00006
2016-01-29    0.00002
2016-02-01    0.00002
2016-02-02    0.00004
2016-02-05    0.00004
2016-02-07    0.00002
2016-02-08    0.00002
2016-02-09    0.00004
2016-02-11    0.00002
2016-02-12    0.00006
2016-02-14    0.00004
2016-02-16    0.00002
2016-02-17    0.00002
2016-02-18    0.00004
2016-02-19    0.00006
2016-02-20    0.00004
2016-02-21    0.00006
2016-02-22    0.00002
2016-02-23    0.00008
2016-02-24    0.00004
2016-02-25    0.00006
2016-02-26    0.00004
2016-02-27    0.00012
2016-02-28    0.00020
2016-02-29    0.00016
2016-03-01    0.00010
2016-03-02    0.00010
2016-03-03    0.00086
2016-03-04    0.00144
2016-03-05    0.02304
2016-03-06    0.01512
2016-03-07    0.03474

##### The ads were created from June 2015 to April 2016. From March 2016, more ads started to be created per day, with percentage jumping from 0.002% to 3%.

In [28]:
# Distribution of last_seen
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

last_seen
2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00760
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02382
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02086
2016-03-29    0.02234
2016-03-30    0.02484
2016-03-31    0.02384
2016-04-01    0.02310
2016-04-02    0.02490
2016-04-03    0.02536
2016-04-04    0.02462
2016-04-05    0.12428
2016-04-06    0.22100
2016-04-07    0.13092
Name: proportion, dtype: float64

##### The ads were last seen for all cars between March and April of 2016. Each day of the distribution constitutes about 2% of all cars.

In [29]:
autos["registration_year"].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

##### On average, cars were registered in 2005. 75% of the cars were registered before 2008. There is at least 2 outliers, as the minimum value is 1000 and the maximum value is 9999, which doesn't make sense.

##### Cars were only invented after 1900, and the latest year in which the ads were created is 2016. So any registration year outside this doesn't make sense and hence can be removed.

In [30]:
autos[autos["registration_year"].between(1900, 2016)]["registration_year"].value_counts(normalize=True, dropna=False).sort_index()

registration_year
1910    0.000187
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000042
1937    0.000083
1938    0.000021
1939    0.000021
1941    0.000042
1943    0.000021
1948    0.000021
1950    0.000062
1951    0.000042
1952    0.000021
1953    0.000021
1954    0.000042
1955    0.000042
1956    0.000104
1957    0.000042
1958    0.000083
1959    0.000146
1960    0.000708
1961    0.000125
1962    0.000083
1963    0.000187
1964    0.000250
1965    0.000354
1966    0.000458
1967    0.000562
1968    0.000541
1969    0.000396
1970    0.000937
1971    0.000562
1972    0.000729
1973    0.000541
1974    0.000500
1975    0.000396
1976    0.000562
1977    0.000458
1978    0.000979
1979    0.000729
1980    0.002020
1981    0.000645
1982    0.000895
1983    0.001104
1984    0.001104
1985    0.002186
1986    0.001582
1987    0.001562
1988    0.002957
1989    0.003769
1990    0.008224
1991    0.007412
1992    0.008141
1993    0.009265
1994    0.013742
1995    0.027338
1996    0.030

##### The perentage of cars registered in each year changed from 0.08% in 1982 to 1% in 1983. Most cars were registered from this year onwards.

In [66]:
# Explore brand column
top20_brands = list(autos["brand"].value_counts(normalize=True, dropna=False).head(20).index)

In [67]:
top20_brands

['volkswagen',
 'opel',
 'bmw',
 'mercedes_benz',
 'audi',
 'ford',
 'renault',
 'peugeot',
 'fiat',
 'seat',
 'skoda',
 'mazda',
 'nissan',
 'smart',
 'citroen',
 'toyota',
 'sonstige_autos',
 'hyundai',
 'volvo',
 'mini']

##### I have chosen to aggregate on the top 20 brands.

In [68]:
mean_price_dict = {}
for brand in top20_brands:
    selected_rows = autos[autos["brand"] == brand]
    mean = (selected_rows["price"].sum()) / len(selected_rows["price"])
    mean_price_dict[brand] = mean

In [69]:
mean_price_dict

{'volkswagen': np.float64(6384.167399644428),
 'opel': np.float64(5106.092657022524),
 'bmw': np.float64(8252.918953766808),
 'mercedes_benz': np.float64(29511.955428812842),
 'audi': np.float64(8965.560354891431),
 'ford': np.float64(7105.662546708824),
 'renault': np.float64(2351.301996672213),
 'peugeot': np.float64(3010.8688186813188),
 'fiat': np.float64(12134.20642201835),
 'seat': np.float64(4219.431455897981),
 'skoda': np.float64(6305.044529262086),
 'mazda': np.float64(3962.542932628798),
 'nissan': np.float64(4588.879310344828),
 'smart': np.float64(3482.971469329529),
 'citroen': np.float64(42657.46362339515),
 'toyota': np.float64(5097.941653160454),
 'sonstige_autos': np.float64(38300.84065934066),
 'hyundai': np.float64(5316.754098360656),
 'volvo': np.float64(31689.908096280087),
 'mini': np.float64(10392.393867924528)}

In [70]:
dictvalues_list = list(mean_price_dict.values())

In [72]:
dictvalues_list.sort()
print(dictvalues_list)

[np.float64(2351.301996672213), np.float64(3010.8688186813188), np.float64(3482.971469329529), np.float64(3962.542932628798), np.float64(4219.431455897981), np.float64(4588.879310344828), np.float64(5097.941653160454), np.float64(5106.092657022524), np.float64(5316.754098360656), np.float64(6305.044529262086), np.float64(6384.167399644428), np.float64(7105.662546708824), np.float64(8252.918953766808), np.float64(8965.560354891431), np.float64(10392.393867924528), np.float64(12134.20642201835), np.float64(29511.955428812842), np.float64(31689.908096280087), np.float64(38300.84065934066), np.float64(42657.46362339515)]


##### Within the the top 20 brands, brand with highest mean price is citroen and lowest mean price is renault.

In [77]:
# Mean mileage for the top 6 brands
top6_mean_mileage_dict = {}
top6_mean_price_dict = {}
for brand in top20_brands[:6]:
    selected_rows = autos[autos["brand"] == brand]
    mean_price = (selected_rows["price"].sum()) / len(selected_rows["price"])
    top6_mean_price_dict[brand] = mean_price
    mean_mileage = (selected_rows["odometer_km"].sum()) / len(selected_rows["odometer_km"])
    top6_mean_mileage_dict[brand] = mean_mileage    

In [81]:
mean_mileage_series = pd.Series(top6_mean_mileage_dict)
print(mean_mileage_series)

volkswagen       128955.272761
opel             129298.663248
bmw              132521.643028
mercedes_benz    130886.142797
audi             129643.941163
ford             124131.934464
dtype: float64


In [84]:
mean_price_series = pd.Series(top6_mean_price_dict)
print(mean_price_series)

volkswagen        6384.167400
opel              5106.092657
bmw               8252.918954
mercedes_benz    29511.955429
audi              8965.560355
ford              7105.662547
dtype: float64


In [87]:
top6_df = pd.DataFrame(mean_price_series, columns=["mean_price"])
top6_df

mean_price
volkswagen      6384.167400
opel            5106.092657
bmw             8252.918954
mercedes_benz  29511.955429
audi            8965.560355
ford            7105.662547

In [90]:
top6_df["mean_mileage"] = mean_mileage_series
top6_df

mean_price   mean_mileage
volkswagen      6384.167400  128955.272761
opel            5106.092657  129298.663248
bmw             8252.918954  132521.643028
mercedes_benz  29511.955429  130886.142797
audi            8965.560355  129643.941163
ford            7105.662547  124131.934464

In [92]:
top6_df.sort_values(by="mean_price", axis=0)

mean_price   mean_mileage
opel            5106.092657  129298.663248
volkswagen      6384.167400  128955.272761
ford            7105.662547  124131.934464
bmw             8252.918954  132521.643028
audi            8965.560355  129643.941163
mercedes_benz  29511.955429  130886.142797

##### There does not seem to be a link between mean price amd mileage. If you look at the top 3 most expensive brands, one would expect as the mean price increases, there would be a decrease in the mean mileage, but that's not the case. FOr th